logpuzzle.py

```diff
@@ -29,7 +32,12 @@ def read_urls(filename):
     Screens out duplicate urls and returns the urls sorted into
     increasing order."""
     # +++your code here+++
-    pass
+    head, _, tail = filename.partition("_")
+    filename_re = re.compile(r'GET\s/(\S*puzzle\S*)\s')
+    with open(filename, 'rt') as f_obj:
+        f_contents = f_obj.read()
+        matches = filename_re.findall(f_contents)
+    return [os.path.join('http://', tail, match) for match in matches]
 
 
 def download_images(img_urls, dest_dir):
```

Comment:
This function can be refactored into functions for downloading image, writing downloaded file, and updating index.html. Try to keep functions down to a single responsiblity i.e. does one thing.